consensus/src/consensus.rs

```diff
@@ -503,6 +524,48 @@ impl<R: ConsensusRpc> ConsensusStateManager<R> {
         Ok(())
     }
 
+    fn bootstrap_offline(&mut self, checkpoint: &[u8], bootstrap: &mut Bootstrap) -> Result<()> {
```

Comment:
It looks like most of the code in `bootstrap` and `bootstrap_offline` is duplicated. Wouldn't it be better to extract the common code into a single function to help prevent potential bugs in the future?